crates/boundless-cli/src/bin/boundless.rs

```diff
@@ -687,29 +687,62 @@ async fn handle_request_command(cmd: &RequestCommands, client: StandardClient) -
             tracing::info!("Successfully retrieved proof for request 0x{:x}", request_id);
             tracing::info!(
-                "Journal: {} - Seal: {}",
-                serde_json::to_string_pretty(&journal)?,
+                "Fulfillment Data: {} - Seal: {}",
+                serde_json::to_string_pretty(&fulfillment_data)?,
                 serde_json::to_string_pretty(&seal)?
             );
             Ok(())
         }
         RequestCommands::VerifyProof { request_id, image_id } => {
             tracing::info!("Verifying proof for request 0x{:x}", request_id);
-            let (journal, seal) =
-                client.boundless_market.get_request_fulfillment(*request_id).await?;
-            let journal_digest = <[u8; 32]>::from(Journal::new(journal.to_vec()).digest()).into();
+
             let verifier_address = client.deployment.verifier_router_address.context("no address provided for the verifier router; specify a verifier address with --verifier-address")?;
             let verifier = IRiscZeroVerifier::new(verifier_address, client.provider());
-
-            verifier
-                .verify(seal, *image_id, journal_digest)
-                .call()
-                .await
-                .map_err(|_| anyhow::anyhow!("Verification failed"))?;
+            let (fulfillment_data, seal) =
+                client.boundless_market.get_request_fulfillment(*request_id).await?;
+            let (req, _) = client.boundless_market.get_submitted_request(*request_id, None).await?;
+
+            let predicate = Predicate::try_from(req.requirements.predicate)?;
+
+            match (&predicate, fulfillment_data.clone()) {
+                (_, FulfillmentData::ImageIdAndJournal(image_id_from_data, journal)) => {
+                    ensure!(
+                        image_id_from_data == Digest::from(<[u8; 32]>::from(*image_id)),
+                        "Image ID mismatch: expected {:?}, got {:?}",
+                        image_id_from_data,
+                        *image_id
+                    );
+                    let journal_digest =
+                        <[u8; 32]>::from(Journal::new(journal.to_vec()).digest()).into();
+
+                    verifier
+                        .verify(seal, *image_id, journal_digest)
+                        .call()
+                        .await
+                        .map_err(|_| anyhow::anyhow!("Verification failed"))?;
+                }
+                (Predicate::ClaimDigestMatch(claim_digest), FulfillmentData::None) => {
+                    verifier
+                        .verifyIntegrity(Receipt {
+                            seal,
+                            claimDigest: <[u8; 32]>::from(*claim_digest).into(),
+                        })
+                        .call()
+                        .await
+                        .map_err(|_| anyhow::anyhow!("Integrity verification failed"))?;
+                }
```

Comment:
The command interface has kind of painted us into a corner here. This is not right, because we did not actually check the given image ID, and there is no way we can. Given the interface, all we can do here is error since we can't actually fulfill the contract of ensuring it matches the image ID.

Reply:
Make sense!